commit aa5f3db70652ad290c4e423828d89d167b7feca9
Author: Dmitriy Paliy <dmitriy.paliy@nokia.com>
Date:   Tue Mar 22 14:29:07 2011 +0200

    Add release all sessions when adapter is switched off
    
    All sessions should be released when adapter is switched off. Then a new
    RequestSession method call always results in change from power off to
    power on such ensuring operational mode. Otherwise, it is possible to
    end up in adapter state being powered off after RequestSession succeded.
    
    g_slist_free is not called after g_slist_foreach because the list is
    updated using g_slist_remove inside of session_free, which is called for
    each element by g_slist_foreach.

diff --git a/src/adapter.c b/src/adapter.c
index cc4f43e..691b963 100644
--- a/src/adapter.c
+++ b/src/adapter.c
@@ -2507,6 +2507,15 @@ static void set_mode_complete(struct btd_adapter *adapter)
 
 	DBG("");
 
+	/*
+	 * g_slist_free is not called after g_slist_foreach because the list is
+	 * updated using g_slist_remove in session_remove which is called by
+         * session_free, which is called for each element by g_slist_foreach.
+	 */
+	if (adapter->mode == MODE_OFF)
+		g_slist_foreach(adapter->mode_sessions, (GFunc) session_free,
+									NULL);
+
 	if (adapter->pending_mode == NULL)
 		return;
 
